src/Network/GRPC/MQTT/Message/Packet.hs

```diff
@@ -371,4 +405,86 @@ isMissingPackets = do
       Nothing -> pure True
       Just ct -> do
         len <- lengthPacketSet pxs
-        pure (fromIntegral len - 1 < ct)
\ No newline at end of file
+        pure (fromIntegral len - 1 < ct)
+
+-- Unsafe ----------------------------------------------------------------------
+
+fromPacketBuildR ::
+  MonadIO m =>
+  Ptr Word8 ->
+  Int ->
+  Packet ByteString ->
+  m (Ptr Word8, Int)
+fromPacketBuildR buffer unused packet = do
+  let build = Encode.reverseMessageBuilder (wireBuildPacket packet)
+  liftIO (fromBuildR build buffer unused)
+
+encodeBufferOffPacket ::
+  MonadIO m =>
+  Ptr Word8 ->
+  Int ->
+  Packet ByteString ->
+  m ByteString
+encodeBufferOffPacket ptr unused packet = do
+  (ptr', unused') <- fromPacketBuildR ptr unused packet
+  size <- liftIO (readTotal ptr' unused')
+  liftIO (unsafePackCStringLen ptr' size)
+
+withEncodeBuffer :: MonadUnliftIO m => Int -> (Ptr Word8 -> m a) -> m a
+withEncodeBuffer packetSizeLimit k = do
+  let bufferSize = metaDataSize + packetSizeLimit
+  buf <- liftIO (newAlignedPinnedByteArray bufferSize metaDataAlign)
+  stateVar <- newIORef BuildRState{currentBuffer = buf, sealedBuffers = mempty}
+  withStablePtr stateVar \statePtr -> do
+    k =<< liftIO (resetEncodeBuffer buf statePtr)
```

Comment:
It looks like this is the only point that `resetEncodeBuffer` is used, but this is only the initialization. Wouldn't you need to call `resetEncodeBuffer` after publishing each packet?

Reply:
Not necessarily. We reuse the same pointer within the serialization loop here at a line 270:
```
                    serialized <- encodeBufferOffPacket ptr bufferSize packet
```
...but, since that is the case, perhaps `resetEncodeBuffer` should be manually inlined to its call site...

Reply:
Right, the pointer would reset. But doesn't `encodeBufferOffPacket` modify the state that is hidden at the start of the buffer?

Reply:
Sure, but how would that change the value `ptr`?

Reply:
It wouldn't change `ptr`. I don't fully understand how the state is used, so I'm just asking if `encodeBufferOffPacket` modifies the state, and that state isn't reset with `resetEndodeBuffer`, will that modified state impact the subsequent call `encodeBufferOffPacket`?

(It is my understanding that the state is stored in the buffer that that `ptr` points to, so it would persist between calls)

Reply:
FWIW, just writing bytes does not update the metadata in the buffer, only the payload bytes in the buffer.  If you do not care about the payload bytes and there is no allocation, then it should be safe to just adjust your write position.  Note that the pointer to the write position and the information about remaining space MUST be kept in sync, otherwise you lose track of where those metadata are located.